driver-core/src/test/unit/com/mongodb/internal/connection/SingleServerClusterSpecification.groovy

```diff
@@ -198,11 +198,6 @@ class SingleServerClusterSpecification extends Specification {
 
         then:
         1 * listener.clusterOpening { it.clusterId == CLUSTER_ID }
-        1 * listener.clusterDescriptionChanged {
```

Comment:
I'm not clear why we no longer get an event when the cluster is first created.

Reply:
In this test, the TestServer constructor creates a new ServerDescription, but in the call to `TestServer#invalidate`, the same ServerDescription is created and sent to sendNotification. In `SingleServerCluster#serverDescriptionChanged`, publishDescription is called with `shouldPublishChangeEvent=false` since the current and previous server descriptions are the same.

Reply:
OK, I think this got weird because, unlike DefaultServer, TestServer sends its change event synchronously from its constructor.  So it actually happens _before_ the call to publishDescription from SingleServerCluster's constructor.

Because we're not using strict mocking, even though there were actually two cluster description changed events being published, we are just matching against the first one in the assertion, and ignoring the second.

With the change in this patch, we get a single event and it matches

        1 * listener.clusterDescriptionChanged {
            it.clusterId == CLUSTER_ID &&
                    it.previousDescription == initialDescription &&
                    it.newDescription == initialDescription

But that's a pretty weird thing to assert since with a real asynchronous Server implementation that's not what would happen.  But you might as well assert it and add a note indicating why that's happening.

We could update TestServer to make it more realistic (by not calling invalidate from its constructor) but let's not do that now.  Can you open a ticket for it?

Reply:
I opened the ticket JAVA-3629 to address this.